Intermittent Unit Test failure - StartupTypeTelemetryTest.WHEN record is called THEN telemetry is recorded with the appropriate label

### Test Run:
Test failed in https://github.com/mozilla-mobile/fenix/pull/26378 - https://firefoxci.taskcluster-artifacts.net/OIThuwo1TRSMqXOucAYXqg/0/public/reports/test/testDebugUnitTest/index.html 

### Stacktrace:
> java.lang.AssertionError: expected:<1> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.mozilla.fenix.perf.StartupTypeTelemetryTest.WHEN record is called THEN telemetry is recorded with the appropriate label(StartupTypeTelemetryTest.kt:96)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:591)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:274)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:88)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
	Suppressed: org.robolectric.android.internal.AndroidTestEnvironment$UnExecutedRunnablesException: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(Looper.getMainLooper()).idle() call.

### Build: 105

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21233)

--- a/app/src/main/java/org/mozilla/fenix/perf/StartupTypeTelemetry.kt
+++ b/app/src/main/java/org/mozilla/fenix/perf/StartupTypeTelemetry.kt
@@ -10,6 +10,7 @@ import androidx.annotation.VisibleForTesting.PRIVATE
 import androidx.lifecycle.DefaultLifecycleObserver
 import androidx.lifecycle.Lifecycle
 import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.DelicateCoroutinesApi
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.GlobalScope
@@ -67,14 +68,19 @@ class StartupTypeTelemetry(
     @VisibleForTesting(otherwise = NONE)
     fun getTestCallbacks() = StartupTypeLifecycleObserver()
 
+    /**
+     * Record startup telemetry based on the available [startupStateProvider] and [startupPathProvider].
+     *
+     * @param dispatcher used to control the thread on which telemetry will be recorded. Defaults to [Dispatchers.IO].
+     */
     @VisibleForTesting(otherwise = PRIVATE)
-    fun record() {
+    fun record(dispatcher: CoroutineDispatcher = Dispatchers.IO) {
         val startupState = startupStateProvider.getStartupStateForStartedActivity(activityClass)
         val startupPath = startupPathProvider.startupPathForActivity
         val label = getTelemetryLabel(startupState, startupPath)
 
         @OptIn(DelicateCoroutinesApi::class)
-        GlobalScope.launch(Dispatchers.IO) {
+        GlobalScope.launch(dispatcher) {
             PerfStartup.startupType[label].add(1)
             logger.info("Recorded start up: $label")
         }

--- a/app/src/test/java/org/mozilla/fenix/perf/StartupTypeTelemetryTest.kt
+++ b/app/src/test/java/org/mozilla/fenix/perf/StartupTypeTelemetryTest.kt
@@ -12,8 +12,11 @@ import io.mockk.impl.annotations.MockK
 import io.mockk.mockk
 import io.mockk.spyk
 import io.mockk.verify
+import kotlinx.coroutines.test.advanceUntilIdle
 import mozilla.components.support.ktx.kotlin.crossProduct
 import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.test.rule.MainCoroutineRule
+import mozilla.components.support.test.rule.runTestOnMain
 import mozilla.telemetry.glean.testing.GleanTestRule
 import org.junit.Assert.assertEquals
 import org.junit.Assert.assertNull
@@ -39,6 +42,9 @@ private val activityClass = HomeActivity::class.java
 class StartupTypeTelemetryTest {
 
     @get:Rule
+    val coroutinesTestRule = MainCoroutineRule()
+
+    @get:Rule
     val gleanTestRule = GleanTestRule(testContext)
 
     private lateinit var telemetry: StartupTypeTelemetry
@@ -62,7 +68,7 @@ class StartupTypeTelemetryTest {
     }
 
     @Test
-    fun `GIVEN all possible path and state combinations WHEN record telemetry THEN the labels are incremented the appropriate number of times`() {
+    fun `GIVEN all possible path and state combinations WHEN record telemetry THEN the labels are incremented the appropriate number of times`() = runTestOnMain {
         val allPossibleInputArgs = StartupState.values().toList().crossProduct(
             StartupPath.values().toList()
         ) { state, path ->
@@ -73,7 +79,8 @@ class StartupTypeTelemetryTest {
             every { stateProvider.getStartupStateForStartedActivity(activityClass) } returns state
             every { pathProvider.startupPathForActivity } returns path
 
-            telemetry.record()
+            telemetry.record(coroutinesTestRule.testDispatcher)
+            advanceUntilIdle()
         }
 
         validTelemetryLabels.forEach { label ->
@@ -87,11 +94,12 @@ class StartupTypeTelemetryTest {
     }
 
     @Test
-    fun `WHEN record is called THEN telemetry is recorded with the appropriate label`() {
+    fun `WHEN record is called THEN telemetry is recorded with the appropriate label`() = runTestOnMain {
         every { stateProvider.getStartupStateForStartedActivity(activityClass) } returns StartupState.COLD
         every { pathProvider.startupPathForActivity } returns StartupPath.MAIN
 
-        telemetry.record()
+        telemetry.record(coroutinesTestRule.testDispatcher)
+        advanceUntilIdle()
 
         assertEquals(1, PerfStartup.startupType["cold_main"].testGetValue())
     }
@@ -99,33 +107,33 @@ class StartupTypeTelemetryTest {
     @Test
     fun `GIVEN the activity is launched WHEN onResume is called THEN we record the telemetry`() {
         launchApp()
-        verify(exactly = 1) { telemetry.record() }
+        verify(exactly = 1) { telemetry.record(any()) }
     }
 
     @Test
     fun `GIVEN the activity is launched WHEN the activity is paused and resumed THEN record is not called`() {
         // This part of the test duplicates another test but it's needed to initialize the state of this test.
         launchApp()
-        verify(exactly = 1) { telemetry.record() }
+        verify(exactly = 1) { telemetry.record(any()) }
 
         callbacks.onPause(mockk())
         callbacks.onResume(mockk())
 
-        verify(exactly = 1) { telemetry.record() } // i.e. this shouldn't be called again.
+        verify(exactly = 1) { telemetry.record(any()) } // i.e. this shouldn't be called again.
     }
 
     @Test
     fun `GIVEN the activity is launched WHEN the activity is stopped and resumed THEN record is called again`() {
         // This part of the test duplicates another test but it's needed to initialize the state of this test.
         launchApp()
-        verify(exactly = 1) { telemetry.record() }
+        verify(exactly = 1) { telemetry.record(any()) }
 
         callbacks.onPause(mockk())
         callbacks.onStop(mockk())
         callbacks.onStart(mockk())
         callbacks.onResume(mockk())
 
-        verify(exactly = 2) { telemetry.record() } // i.e. this should be called again.
+        verify(exactly = 2) { telemetry.record(any()) } // i.e. this should be called again.
     }
 
     private fun launchApp() {

